src/utils/proxyMap.ts

```diff
@@ -0,0 +1,113 @@
+import { proxy } from 'valtio'
+
+type KeyValRecord<K, V> = {
+  key: K
+  value: V
+}
+
+type InternalProxyMap<K, V> = Map<K, V> & {
+  data: KeyValRecord<K, V>[]
+  toJSON: object
+}
+
+/**
+ * proxyMap
+ *
+ * This is to create a proxy which mimic the native Map behavior.
+ * The API is the same as Map API
+ *
+ * @example
+ * import { proxyMap } from 'valtio/utils'
+ * const state = proxyMap([["key", "value"]])
+ *
+ * //can be used inside a proxy as well
+ * const state = proxy({
+ *   count: 1,
+ *   map: proxyMap()
+ * })
+ */
+export const proxyMap = <K, V>(
+  entries: Iterable<readonly [K, V]> | null = []
+): Map<K, V> => {
+  const map = proxy<InternalProxyMap<K, V>>({
+    data:
+      entries === null
+        ? []
+        : Array.from(entries).map((v) => ({ key: v[0], value: v[1] })),
+    has(key) {
+      return this.data.findIndex((p) => p.key === key) !== -1
+    },
+    set(key, value) {
+      const idx = this.data.findIndex((p) => p.key === key)
+      const payload = {
+        key,
+        value,
+      }
+      if (idx !== -1) {
+        this.data[idx] = payload
```

Comment:
For several reasons (one thing I notice is the case if the value is the same), we should not replace the existing object.
```suggestion
        this.data[idx].value = value
```
In this case, it's probably simpler to use `.find` instead of `.findIndex` in L41.